lightning/src/util/persist.rs

```diff
@@ -205,3 +213,760 @@ where
+	}
+
+	/// Deserialize a channel monitor.
+	fn deserialize_monitor<ES: Deref, SP: Deref>(
+		&self,
+		entropy_source: ES,
+		signer_provider: SP,
+		monitor_name: &MonitorName,
+	) -> io::Result<(
+		BlockHash,
+		ChannelMonitor<<SP::Target as SignerProvider>::Signer>,
+	)>
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized,
+	{
+		let key = monitor_name.as_str();
+		let outpoint: OutPoint = monitor_name.try_into()?;
+		let mut monitor_cursor = io::Cursor::new(
+			self.kv_store
+				.read(CHANNEL_MONITOR_PERSISTENCE_NAMESPACE, key)?,
+		);
+		// Discard the sentinel bytes if found.
+		if monitor_cursor
+			.get_ref()
+			.starts_with(MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL)
+		{
+			monitor_cursor.set_position(MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL.len() as u64);
+		}
+		match <(
+			BlockHash,
+			ChannelMonitor<<SP::Target as SignerProvider>::Signer>,
+		)>::read(&mut monitor_cursor, (&*entropy_source, &*signer_provider))
+		{
+			Ok((blockhash, channel_monitor)) => {
+				if channel_monitor.get_funding_txo().0.txid != outpoint.txid
+					|| channel_monitor.get_funding_txo().0.index != outpoint.index
+				{
+					log_error!(
+						self.logger,
+						"ChannelMonitor was stored under the wrong key! context: {}",
+						monitor_name.as_str()
+					);
+					return Err(io::Error::new(
+						io::ErrorKind::InvalidData,
+						"ChannelMonitor was stored under the wrong key",
+					));
+				}
+				Ok((blockhash, channel_monitor))
+			}
+			Err(e) => {
+				log_error!(
+					self.logger,
+					"Failed to deserialize ChannelMonitor! reason: {} context: {}",
+					e,
+					monitor_name.as_str()
+				);
+				Err(io::Error::new(
+					io::ErrorKind::InvalidData,
+					"Failed to deserialize ChannelMonitor",
+				))
+			}
+		}
+	}
+
+	/// Deserialize a channel monitor update.
+	fn deserialize_monitor_update(
+		&self,
+		monitor_name: &MonitorName,
+		update_name: &UpdateName,
+	) -> io::Result<ChannelMonitorUpdate> {
+		let ns = self.monitor_update_namespace(monitor_name);
+		let key = update_name.as_str();
+		let update_bytes = self.kv_store.read(&ns, key)?;
+		match ChannelMonitorUpdate::read(&mut io::Cursor::new(update_bytes)) {
+			Ok(update) => Ok(update),
+			Err(e) => {
+				log_error!(
+					self.logger,
+					"update could note be decoded! reason: {} context: {}",
+					e,
+					update_name.as_str()
+				);
+				Err(io::Error::new(
+					io::ErrorKind::InvalidData,
+					"update decode failed",
+				))
+			}
+		}
+	}
+
+	/// Delete updates with an update_id lower than the given channel monitor.
+	fn delete_stale_updates<ChannelSigner: WriteableEcdsaChannelSigner>(
+		&self,
+		channel_id: OutPoint,
+		monitor: &ChannelMonitor<ChannelSigner>,
+	) -> io::Result<()> {
+		let monitor_name: MonitorName = channel_id.into();
+		let update_names = self.list_update_names(&monitor_name)?;
+		for update_name in update_names {
+			let update = self.deserialize_monitor_update(&monitor_name, &update_name)?;
```

Comment:
Right, this is similar to https://github.com/lightningdevkit/rust-lightning/pull/2359/files#r1310627097 ... I think if our "contract" is that the update name _must_ be a parseable u64 that matches the update_id, and if you get it wrong, it's on you... this is fine. 

I was even more interested in verifying before deleting, since you could ostensibly recover if you somehow messed this up. But I think it's a reasonable contract to say, you can't mess this up (or if you want to be able to, back up updates).

Reply:
Right, this is similar to https://github.com/lightningdevkit/rust-lightning/pull/2359/files#r1310627097 ... I think if our "contract" is that the update name _must_ be a parseable u64 that matches the update_id, and if you get it wrong, it's on you... this is fine. 

I was even more interested in verifying before deleting, since you could ostensibly recover if you somehow messed this up. But I think it's a reasonable contract to say, you can't mess this up (or if you want to be able to, back up updates).

Reply:
> if you really want to make sure and don't trust this implementation

It's not this implementation that I'm not trusting -- it's the chain of stuff that writes and reads it back.

Reply:
Added the debug assertion